compiler/infra/Assert.cpp

```diff
@@ -165,4 +168,151 @@ namespace TR
       va_fatal_assertion(file, line, condition, format, ap);
       va_end(ap);
       }
+
+   static bool containsNode(TR::Node *node, TR::Node *target, TR::NodeChecklist& nodeChecklist)
+      {
+      if (node == target)
+         return true;
+
+      if (nodeChecklist.contains(node))
+         return false;
+
+      nodeChecklist.add(node);
+
+      for (int i = 0; i < node->getNumChildren(); i++)
+         {
+         if (containsNode(node->getChild(i), target, nodeChecklist))
+            return true;
+         }
+
+      return false;
+      }
+
+   static void markInChecklist(TR::Node *node, TR_BitVector &nodeChecklist)
+      {
+      if (nodeChecklist.isSet(node->getGlobalIndex()))
+         return;
+
+      nodeChecklist.set(node->getGlobalIndex());
+
+      for (int i = 0; i < node->getNumChildren(); i++)
+         markInChecklist(node->getChild(i), nodeChecklist);
+      }
+
+   void NodeAssertionContext::printContext() const
+      {
+      if (!_node) return;
+
+      static bool printFullContext = feGetEnv("TR_AssertFullContext");
+      TR::Compilation *comp = TR::comp();
+      TR_Debug *debug = comp->findOrCreateDebug();
+
+      fprintf(stderr, "\nNode context:\n\n");
+
+      if (printFullContext)
+         {
+         debug->printIRTrees(TR::IO::Stderr, "Assertion Context", comp->getMethodSymbol());
+         debug->print(TR::IO::Stderr, comp->getMethodSymbol()->getFlowGraph());
+         if (comp->getKnownObjectTable())
+            comp->getKnownObjectTable()->dumpTo(TR::IO::Stderr, comp);
+         }
+      else
+         {
+         fprintf(stderr, "...\n");
+
+         TR::NodeChecklist checkedNodeChecklist(comp);
+
+         TR_BitVector commonedNodeChecklist;
+         commonedNodeChecklist.init(0, comp->trMemory(), heapAlloc, growable);
+
+         bool foundNode = false;
+         for (TR::TreeTopIterator it(comp->getMethodSymbol()->getFirstTreeTop(), comp); it != NULL; ++it)
```

Comment:
It might make sense to use getStartTree here.

Reply:
Good point. Changed in eb741f025f91e67c92e694cec054c3d4a69087aa.